lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+`False`, here `h`, and concludes whatever is the goal.
+-/
+theorem nonsense (h : False) : 0 = 1 := False.elim h
+
+/-
+The `contradiction` tactic uses any `False` in the hypothesis to conclude the
+goal.
+-/
+theorem more_nonsense (h : False) : 1 = 2 := by contradiction
+
+/-
+To illustrate constructive vs classical logic, we now prove the contrapositive
+theorem.
+
+The forward direction does not require classical logic.
+-/
+theorem contrapositive_forward' : (p → q) → (¬q → ¬p) :=
+  fun pq => fun hqf => fun hp => hqf (pq hp)
+/-
+Use the definition `¬q := q → False`. Notice that we have to construct `p →
+False` given `p → q` and `q → False`. This is just function composition.
+-/
+
+/-
+The above proof, using tactics.
+-/
+theorem contrapositive_forward : (p → q) → (¬q → ¬p) := by
+  intro hpq
+  intro
+  intro hp
+  specialize hpq hp
+  contradiction
+
+/-
+The reverse requires classical logic.
+
+Here, we are required to construct a `q` given values of following types:
+
+  - `(q → False) → (p → False)`.
+  - `p`.
+
+This is impossible without using the law of excluded middle.
+-/
+theorem contrapositive_reverse' : (¬q → ¬p) → (p → q) :=
+  fun hnqnp =>
+  Classical.byCases
+    (fun hq  => fun  _ => hq)
+    (fun hnq => fun hp => absurd hp (hnqnp hnq))
+/-
+Law of excluded middle tells us that we will have a `q` or a `q → False`. In the
+first case, it is trivial to construct a `q`, we already have it. In the second
+case, we give the `q → False` to obtain a `p → False`.  Then, we use the fact
+(in constructive logic) that given `p` and `p → False`, we can construct
+`False`. Once, we have `False`, we can construct anything, and specifically `q`.
+-/
+
+/-
+Same proof, using tactics.
+-/
+theorem contrapositive_reverse : (¬q → ¬p) → (p → q) := by
+  intro hnqnp
+  intro hp
+  have emq := Classical.em q
+  cases emq
+  case inl _ => assumption
+  case inr h => specialize hnqnp h ; contradiction
+
+/-
+To illustrate how we can define an work with axiomatic systems. Here is a
+definition of Groups and some proofs directly translated from "Topics in
+Algebra" by Herstein, Second edition.
+-/
+
+/-
+A class defines a set of axioms
+-/
+class Group (G : Type u) where
+  op : G → G → G
+  assoc : ∀ a b c : G, op (op a b) c = op a (op b c)
+  e : G
+  identity: ∀ a : G, op a e = a ∧ op e a = a
+  inverse: ∀ a : G, ∃ b : G, op a b = e ∧ op b a = e
+
+/-
+Let us introduce some notation to make this convenient.
+-/
+open Group
+infixl:70 " * " => op
+
+/-
+`G` will always stand for a group. Variables `a b c` will be elements of that
+group.
+-/
+variable [Group G] {a b c : G}
+
+def is_identity (e' : G) := ∀ a : G, (a * e' = a ∧ e' * a = a)
+
+/-
+We prove that the identity element is unique.
+-/
+theorem identity_element_unique : ∀ e' : G, is_identity e' → e' = e := by
+  intro e'
+  intro h
+  specialize h e
+  obtain ⟨h1, _⟩ := h
+  have h' := identity e'
+  obtain ⟨_, h2⟩ := h'
```

Comment:
Why have we started using `obtain` instead of the `let` up until now?

Reply:
Are `obtain` and `let` interchangeable?

Reply:
They're not. `obtain` is a `have` + full expressivity of what you call 'rcases patterns'. `let` is a proof-relevant binder that just happens to be able to deconstruct some simple patterns.

E.g. the former can do things like:
```have X : True ∨ False := sorry
  obtain h₁ | h₂ := X```
  
I would say `obtain` is almost a superset but that's somewhat misleading because `let's` are meant to be used for a different purpose + there's the distinction of whether they keep the data on their rhs around.

So, just think of them as two different features. `obtain` allows you to inspect terms, `let` allows you to introduce new terms.

Reply:
I see. I have replaced all `let` with `have` in proofs. I think in proofs this is the right thing to do due to proof irrelevance as you mentioned.

Thanks for your comments. They have been very helpful.